### PR TITLE
Add CI Pool vmss

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

Fixes [VSTS 7820180](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/7820180)

### What this PR does / why we need it:

Introduces new VMSS into env for CI testing

### Test plan for issue:

Test in prod

### Is there any documentation that needs to be updated for this PR?

TBD

### CODE REVIEW

**Comment by mjudeikis** [comment]
Still missing bits. Can't test as lacking some accesses.

File: pkg/deploy/generator/resources.go [hunk 1]
```diff
@@ -1557,3 +1563,195 @@ func (g *generator) acrRbac() []*arm.Resource {
...
+mkdir -p /usr/local/go
+tar -C /usr/local/go -xzf /tmp/go${GO_VERSION}.linux-amd64.tar.gz --strip-components=1 go
+ln -s /usr/local/go/bin/* /usr/bin/
+echo "GOROOT=/usr/local/go" | tee -a /etc/environment
+
+# configure agent
+AGENTNAME=$AGENTPREFIX-$HOSTNAME
+sudo -u cloud-user ./config.sh --unattended --url https://dev.azure.com/msazure --auth pat --token "$AZPTOKEN" --pool "$POOLNAME" --agent "$AGENTNAME"
+
+# default svc names are genered with random artifacts and not intuitive.
```
Comment (nilsanderselde) [inline pkg/deploy/generator/resources.go:1627]:
```suggestion
# default svc names are generated with random artifacts and not intuitive.
```

File: .pipelines/ci.yml [hunk 2]
```diff
@@ -12,33 +12,25 @@ jobs:
   strategy:
     matrix:
       Python36:
-        python.version: '3.6'
+        python.version: '36'
```
Comment (mjudeikis) [inline .pipelines/ci.yml:15]: TODO: fix this

File: pkg/deploy/generator/resources.go [hunk 3]
```diff
@@ -1557,3 +1563,176 @@ func (g *generator) acrRbac() []*arm.Resource {
...
+EOF
+
+# install common tooling
+yum -y install docker azure-cli jq git python36 python3-pip jq gcc \
+gpgme-devel glib2-devel make libseccomp libseccomp-devel glibc-static libassuan libassuan-devel
+
+pip3 install virtualenv
+
+# install golang
+GO_VERSION=1.13
```
Comment (hawkowl) [inline pkg/deploy/generator/resources.go:1603]: This will install Go 1.13.0, not the latest patch version of 1.13 (which right now is 1.13.15)

File: pkg/deploy/generator/resources.go [hunk 4]
```diff
@@ -1557,3 +1563,176 @@ func (g *generator) acrRbac() []*arm.Resource {
...
+curl -sL https://dl.google.com/go/go${GO_VERSION}.linux-amd64.tar.gz -o /tmp/go${GO_VERSION}.linux-amd64.tar.gz
+mkdir -p /usr/local/go
+tar -C /usr/local/go -xzf /tmp/go${GO_VERSION}.linux-amd64.tar.gz --strip-components=1 go
+ln -s /usr/local/go/bin/* /usr/bin/
+echo "GOROOT=/usr/local/go" | tee -a /etc/environment
+
+# Install azp agent
+wget https://vstsagentpackage.azureedge.net/agent/2.172.2/vsts-agent-linux-x64-2.172.2.tar.gz -O /tmp/vsts-agent-linux-x64-2.172.2.tar.gz
```
Comment (hawkowl) [inline pkg/deploy/generator/resources.go:1611]: Should this be templated so that we can update one variable to update the VSTS agent? They've also updated it, current version is now v2.173.0.

**Comment by mjudeikis** [comment]
I expect some pipelines to be broken after this merged... hard to test

File: hack/devtools/deploy-shared-env.sh [hunk 5]
```diff
@@ -56,12 +60,16 @@ deploy_env_dev_override() {
         -n env-development \
         --template-file deploy/env-development.json \
         --parameters \
+            "azpToken=$AZPTOKEN" \
```
Comment (mjudeikis) [inline hack/devtools/deploy-shared-env.sh:63]: This will move to keyvault once we get access to it....

**Comment by kwoodson** [comment]
@mjudeikis Thanks for this work.  Let's get this merged and we can improve on this process.

**Comment by kwoodson** [comment]
@jim-minter @asalkeld PTAL

**Comment by mjudeikis** [comment]
CI pipeline for this to be "re-deployable" - https://msazure.visualstudio.com/AzureRedHatOpenShift/_build?definitionId=161268&_a=summary

File: .pipelines/vars.yml [hunk 6]
```diff
@@ -1,5 +1,3 @@
 variables:
-  GOBIN: $(GOPATH)/bin   # Go binaries path
```
Comment (jim-minter) [inline .pipelines/vars.yml:2]: this seems to still be referenced?

File: .pipelines/templates/template-init-env.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+parameters:
+  modulePath: ''
+steps:
+- script: |
+    set +e
+    # this file should go away once moved to go modules
+    mkdir -p '${{ parameters.modulePath }}'
+    ls -A | grep -v gopath | xargs -t -I '{}' mv {} ${{ parameters.modulePath }}
```
Comment (jim-minter) [inline .pipelines/templates/template-init-env.yml:8]: still pretty ugly, wish there was a better way :-|

File: .pipelines/deploy-dev-env.yml [hunk 8]
```diff
@@ -32,3 +26,6 @@ jobs:
   - template: ./templates/template-deploy-shared-env.yml
     parameters:
       workingDirectory: ${{ variables.modulePath }}
+      subscriptionID: $(AZURE_SUBSCRIPTION_ID)
+      azureDevOpsJSONSPN: $(aro-v4-e2e-devops-spn)
+      azpToken: $(AZPTOKEN)
```
Comment (jim-minter) [inline .pipelines/deploy-dev-env.yml:31]: needs to be moved away

File: deploy/env-development.json [hunk 9]
```diff
@@ -117,6 +135,103 @@
...
+                                ]
+                            }
+                        }
+                    },
+                    "storageProfile": {
+                        "imageReference": {
+                            "publisher": "RedHat",
+                            "offer": "RHEL",
+                            "sku": "8.1",
+                            "version": "8.1.2020020415"
```
Comment (jim-minter) [inline deploy/env-development.json:169]: I'm uncomfortable that we're using RHEL 8 here and 7 elsewhere, it would be better if we could standardise...

Comment (mjudeikis) [reply deploy/env-development.json:169]: The issue is with `libseccomp libseccomp-devel` pkg state in rhel7. How much of priority it is as it might take ~1 day for a fix.

Comment (mjudeikis) [reply deploy/env-development.json]: Done

File: docs/prepare-a-shared-rp-development-environment.md [hunk 10]
```diff
@@ -102,8 +102,8 @@ locations.
    utility:
 
    ```bash
-   go run ./hack/genkey -client firstparty-development
-   mv firstparty-development.* secrets
+   go run ./hack/genkey -client firstparty
```
Comment (mjudeikis) [inline docs/prepare-a-shared-rp-development-environment.md:105]: Align the names with what is provisioned in production/Int...

File: deploy/env-development.json [hunk 11]
```diff
@@ -117,6 +135,103 @@
                 "[resourceId('Microsoft.Network/virtualNetworks', 'dev-vnet')]"
             ]
         },
+        {
```
Comment (jim-minter) [inline deploy/env-development.json:138]: do we want to put the vmss in a separate template?  If we don't it means we'll get CI VMSSes when we deploy to our dev subscription?

File: docs/prepare-a-shared-rp-development-environment.md (hunk 10 above)
Comment (jim-minter) [reply docs/prepare-a-shared-rp-development-environment.md:105]: please manually update our shared secrets when this merges?

File: pkg/deploy/generator/resources.go [hunk 12]
```diff
@@ -1573,3 +1579,197 @@ func (g *generator) rpVersionStorageAccount() *arm.Resource {
...
+fi
+EOF
+
+# install golang
+GO_VERSION=1.13.15
+curl -sL https://dl.google.com/go/go${GO_VERSION}.linux-amd64.tar.gz -o /tmp/go${GO_VERSION}.linux-amd64.tar.gz
+mkdir -p /usr/local/go
+tar -C /usr/local/go -xzf /tmp/go${GO_VERSION}.linux-amd64.tar.gz --strip-components=1 go
+ln -s /usr/local/go/bin/* /usr/bin/
+echo "GOROOT=/usr/local/go" | tee -a /etc/environment
```
Comment (jim-minter) [inline pkg/deploy/generator/resources.go:1639]: https://dave.cheney.net/2013/06/14/you-dont-need-to-set-goroot-really

File: pkg/deploy/generator/resources.go [hunk 13]
```diff
@@ -1573,3 +1579,197 @@ func (g *generator) rpVersionStorageAccount() *arm.Resource {
...
+
+# Install azp agent
+VSTS_AGENT=2.172.2
+wget https://vstsagentpackage.azureedge.net/agent/${VSTS_AGENT}/vsts-agent-linux-x64-${VSTS_AGENT}.tar.gz -O /tmp/vsts-agent-linux-x64-${VSTS_AGENT}.tar.gz
```
Comment (jim-minter) [inline pkg/deploy/generator/resources.go:1643]: standardise on curl, not wget

File: pkg/deploy/generator/resources.go [hunk 14]
```diff
@@ -1573,3 +1579,197 @@ func (g *generator) rpVersionStorageAccount() *arm.Resource {
...
+cat > /home/cloud-user/agent/.path <<'EOF'
+/opt/rh/rh-python36/root/usr/bin:/opt/rh/rh-git29/root/usr/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/cloud-user/.local/bin:/home/cloud-user/bin
+EOF
+
+cat > /home/cloud-user/agent/.env <<'EOF'
+LANG=en_US.UTF-8
+LD_LIBRARY_PATH=/opt/rh/rh-python36/root/usr/lib64:/opt/rh/httpd24/root/usr/lib64
+PERL5LIB=/opt/rh/rh-git29/root/usr/share/perl5/vendor_perl
+HOME=/home/cloud-user
+EOF
```
Comment (jim-minter) [inline pkg/deploy/generator/resources.go:1665]: interested to know why these are needed

File: Makefile [hunk 15]
```diff
@@ -104,14 +104,14 @@ pyenv${PYTHON_VERSION}:
 secrets:
 	@[ "${SECRET_SA_ACCOUNT_NAME}" ] || ( echo ">> SECRET_SA_ACCOUNT_NAME is not set"; exit 1 )
 	rm -rf secrets
-	az storage blob download --auth-mode login -n secrets.tar.gz -c secrets -f secrets.tar.gz --account-name ${SECRET_SA_ACCOUNT_NAME} >/dev/null
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply Makefile:107]: no. It is needed for CI. in Dev it gives a warning, but works.

File: .pipelines/templates/template-push-images-to-acr.yml [hunk 16]
```diff
@@ -8,8 +8,13 @@ steps:
 
     export RP_IMAGE_ACR=${{ parameters.rpImageACR }}
 
-    az acr login --name "$RP_IMAGE_ACR"
     # azure checkouts commit, so removing master reference when publishing image
     export BRANCH=$(Build.SourceBranchName)
+
+    az acr login --name "$RP_IMAGE_ACR" --expose-token > acr.json
+    podman login $(cat acr.json | jq -r .loginServer) -u 00000000-0000-0000-0000-000000000000 -p $(cat acr.json | jq -r .accessToken)
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply .pipelines/templates/template-push-images-to-acr.yml:15]: Rhel7 - docker
Rhel8 - podman and this. gonna rollback

File: deploy/env-development.json (hunk 11 above)
Comment (mjudeikis) [reply deploy/env-development.json:138]: What is the parameter `ReplicasCount` and keep the consistency of the code? Fewer deltas?

File: pkg/deploy/generator/resources.go (hunk 14 above)
Comment (mjudeikis) [reply pkg/deploy/generator/resources.go:1665]: Because MSFT agent runs in the custom context/shell interpreter and does not honour hosts configs for these values. It works on Ubuntu, but they are not friends with rhel :/ 
We ship python36 and git2 (hard requirement for VSTS) only ins RHSCL. So we need to make sure agent knows where to look for them when executing tests in msft shell. On rhel8 it is easier as these are in standard PATH and agent understands it. 

Painful.

Very.

File: hack/devtools/deploy-shared-env.sh [hunk 17]
```diff
@@ -39,12 +39,17 @@ deploy_env_dev() {
         -n env-development \
         --template-file deploy/env-development.json \
         --parameters \
+            "agentPrefix=ARO-RHEL" \
+            "azpToken"="$AZPTOKEN" \
```
Comment (jim-minter) [inline hack/devtools/deploy-shared-env.sh:43]:
```suggestion
            "azpToken=$AZPTOKEN" \
```

File: hack/devtools/deploy-shared-env.sh [hunk 18]
```diff
@@ -39,12 +39,17 @@ deploy_env_dev() {
         -n env-development \
         --template-file deploy/env-development.json \
         --parameters \
+            "agentPrefix=ARO-RHEL" \
```
Comment (jim-minter) [inline hack/devtools/deploy-shared-env.sh:42]:
```suggestion
            "agentPrefix=ARO-RHEL-" \
```

(to correspond with below?)

File: hack/devtools/deploy-shared-env.sh [hunk 19]
```diff
@@ -56,12 +61,17 @@ deploy_env_dev_override() {
         -n env-development \
         --template-file deploy/env-development.json \
         --parameters \
+            "agentPrefix=ARO-RHEL-" \
+            "azpToken"="$AZPTOKEN" \
```
Comment (jim-minter) [inline hack/devtools/deploy-shared-env.sh:65]:
```suggestion
            "azpToken=$AZPTOKEN" \
```

File: hack/e2e/run-rp-and-e2e.sh [hunk 20]
```diff
@@ -168,6 +168,7 @@ clean_e2e() {
 }
 
 export CLUSTER="v4-e2e-V$BUILD_BUILDID"
+export DATABASE_NAME=v4-e2e-V$BUILD_BUILDID
```
Comment (jim-minter) [inline hack/e2e/run-rp-and-e2e.sh:171]:
```suggestion
export DATABASE_NAME="v4-e2e-V$BUILD_BUILDID"
```

File: pkg/deploy/generator/resources.go [hunk 21]
```diff
@@ -1573,3 +1582,204 @@ func (g *generator) rpVersionStorageAccount() *arm.Resource {
...
+VSTS_AGENT=2.172.2
+curl -sL https://vstsagentpackage.azureedge.net/agent/${VSTS_AGENT}/vsts-agent-linux-x64-${VSTS_AGENT}.tar.gz -o /tmp/vsts-agent-linux-x64-${VSTS_AGENT}.tar.gz
+mkdir /home/cloud-user/agent && cd /home/cloud-user/agent
+tar zxvf /tmp/vsts-agent-linux-x64-${VSTS_AGENT}.tar.gz
+./bin/installdependencies.sh
+sudo chown cloud-user:root -R /home/cloud-user/agent
+
+# configure agent
+AGENTNAME=$AGENTPREFIX-$HOSTNAME
+sudo -u cloud-user ./config.sh --unattended --url https://dev.azure.com/msazure --auth pat --token "$AZPTOKEN" --pool "$POOLNAME" --agent "$AGENTNAME"
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/deploy/generator/resources.go:1660]: URL should not change for us. PoolName - might.

File: .pipelines/templates/template-deploy-shared-env.yml [hunk 22]
```diff
@@ -1,12 +1,16 @@
...
+  azureDevOpsJSONSPN: ''
 steps:
 - script: |
     set -e
-    cd ${{ parameters.workingDirectory }}
 
     . secrets/env
+    export DATABASE_NAME=v4-e2e-V$BUILD_BUILDID
     . ./hack/devtools/deploy-shared-env.sh
+    base64 -d >devops-spn.json <<<${{ parameters.azureDevOpsJSONSPN }}
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply .pipelines/templates/template-deploy-shared-env.yml:11]: VSTS 8171837

File: .pipelines/templates/template-deploy-shared-env.yml [hunk 23]
```diff
@@ -1,12 +1,16 @@
...
-  workingDirectory: ''
+  subscriptionID: ''
+  azureDevOpsJSONSPN: ''
 steps:
 - script: |
     set -e
-    cd ${{ parameters.workingDirectory }}
 
     . secrets/env
+    export DATABASE_NAME=v4-e2e-V$BUILD_BUILDID
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply .pipelines/templates/template-deploy-shared-env.yml:9]: Yes, it does not. But this is due to mess in Shell scripts. This will be refactored once we got rid of shell magic :/ 
TL;DR: `deploy-shared-env.sh` and `run-rp-and-e2e.sh` should not share as much copy-paste as they do to avoid something like this...

File: Makefile [hunk 24]
```diff
@@ -1,9 +1,9 @@
 SHELL = /bin/bash
 COMMIT = $(shell git rev-parse --short HEAD)$(shell [[ $$(git status --porcelain) = "" ]] || echo -dirty)
 ARO_IMAGE ?= ${RP_IMAGE_ACR}.azurecr.io/aro:$(COMMIT)
-
+CGO_CFLAGS="-Dgpgme_off_t=off_t"
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply Makefile:4]: `gcc: error: "-Dgpgme_off_t=off_t": No such file or directory` Makefiles....

File: .pipelines/templates/template-deploy-e2e-deps.yml [hunk 25]
```diff
@@ -2,8 +2,6 @@ parameters:
   workingDirectory: ''
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply .pipelines/templates/template-deploy-e2e-deps.yml:2]: bump

File: Makefile (hunk 15 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply Makefile:107]: bump

File: Makefile (hunk 24 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply Makefile:4]: the double quotes are the problem: `export CGO_CFLAGS=-Dgpgme_off_t=off_t` please

**Review by jim-minter — CHANGES_REQUESTED** [review]
Couple of missing changes - please check:

**Review by asalkeld — APPROVED** [review]
